File: src/main/java/net/effize/bandlog/team/application/TeamQueryService.java
package net.effize.bandlog.team.application;

import net.effize.bandlog.team.application.dto.TeamInfoResponse;
import net.effize.bandlog.team.application.dto.TeamsResponse;
import net.effize.bandlog.team.application.converter.MemberRoleConverter;
import net.effize.bandlog.team.domain.model.Member;
import net.effize.bandlog.team.domain.model.Team;
import net.effize.bandlog.team.domain.model.TeamId;
import net.effize.bandlog.identity.api.UserService;
import net.effize.bandlog.identity.api.dto.UserDto;
import net.effize.bandlog.team.domain.model.UserId;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class TeamQueryService {
    private final TeamRepositoryService teamRepositoryService;
    private final UserService userService;
    private final MemberRoleConverter memberRoleConverter;

    public TeamQueryService(TeamRepositoryService teamRepositoryService, UserService userService, MemberRoleConverter memberRoleConverter) {
        this.teamRepositoryService = teamRepositoryService;
        this.userService = userService;
        this.memberRoleConverter = memberRoleConverter;
    }

    public TeamInfoResponse teamInfo(UserId authUserId, Long teamId) {
        Team foundTeam = teamRepositoryService.findTeamById(TeamId.of(teamId));
        List<Member> members = teamRepositoryService.findMembersByTeam(foundTeam);
        long meCount = members.stream().filter(member -> member.userId().equals(authUserId)).count();
        if (meCount <= 0) throw new IllegalStateException("User is not a member of the team");

        List<UserDto> users = userService.findAllByIdIn(
                members
                        .stream()
                        .map(member -> member.userId().value())
                        .toList()
        );

        Map<Long, UserDto> userMap = users.stream()
                .collect(Collectors.toMap(
                        user -> user.id(),
                        user -> user
                ));

        return new TeamInfoResponse(
                foundTeam.name(),
                foundTeam.description(),
                foundTeam.inviteCode(),
                members.stream()
                        .map(member -> new TeamInfoResponse.MemberInfo(
                                userMap.get(member.userId().value()).nickname(),
                                memberRoleConverter.toDto(member.role())
                        ))
                        .toList()
        );
    }

    public TeamsResponse myTeams(UserId authUserId) {
        List<Team> teams = teamRepositoryService.findTeamsByUserId(authUserId);

        return new TeamsResponse(
                teams.stream().map(team -> {
                    List<Member> members = teamRepositoryService.findMembersByTeam(team);

                    List<UserDto> users = userService.findAllByIdIn(
                            members
                                    .stream()
                                    .map(member -> member.userId().value())
                                    .toList()
                    );

                    Map<Long, UserDto> userMap = users.stream()
                            .collect(Collectors.toMap(
                                    user -> user.id(),
                                    user -> user
                            ));

                    return new TeamInfoResponse(
                            team.name(),
                            team.description(),
                            team.inviteCode(),
                            members.stream()
                                    .map(member -> new TeamInfoResponse.MemberInfo(
                                            userMap.get(member.userId().value()).nickname(),
                                            memberRoleConverter.toDto(member.role())
                                    ))
                                    .toList()
                    );
                }).toList()
        );
    }
}

File: src/main/kotlin/net/effize/bandlog/team/application/TeamQueryService.kt
package net.effize.bandlog.team.application

import net.effize.bandlog.common.id.UserId
import net.effize.bandlog.team.adapter.`in`.web.response.TeamInfoResponse
import net.effize.bandlog.team.adapter.`in`.web.response.TeamsResponse
import net.effize.bandlog.team.api.TeamQueryPort
import net.effize.bandlog.team.application.port.UserInfoPort
import net.effize.bandlog.team.domain.MemberId
import net.effize.bandlog.team.domain.MemberRole
import net.effize.bandlog.team.domain.TeamId
import org.springframework.stereotype.Service

@Service
class TeamQueryService(
    private val teamService: TeamService,
    private val userInfoPort: UserInfoPort
) : TeamQueryPort {

    fun teamInfo(authUserId: UserId, teamId: Long): TeamInfoResponse {
        val foundTeam = teamService.activeTeam(TeamId.of(teamId))
        val members = teamService.membersOf(foundTeam)
        val meCount = members.count { it.userId() == authUserId }
        if (meCount <= 0) throw IllegalStateException("User is not a member of the team")

        val users = userInfoPort.findAllByIds(members.map { it.userId() })
        val userMap = users.associateBy { it.id.value }

        return TeamInfoResponse(
            name = foundTeam.name,
            description = foundTeam.description,
            inviteCode = foundTeam.inviteCode,
            members = members.map { member ->
                TeamInfoResponse.MemberInfo(
                    nickname = userMap[member.userId().value]?.nickname ?: "",
                    role = member.role
                )
            }
        )
    }

    fun myTeams(authUserId: UserId): TeamsResponse {
        val teams = teamService.teamsOfUser(authUserId)

        return TeamsResponse(
            teams = teams.map { team ->
                val members = teamService.membersOf(team)
                val users = userInfoPort.findAllByIds(members.map { it.userId() })
                val userMap = users.associateBy { it.id.value }

                TeamInfoResponse(
                    name = team.name,
                    description = team.description,
                    inviteCode = team.inviteCode,
                    members = members.map { member ->
                        TeamInfoResponse.MemberInfo(
                            nickname = userMap[member.userId().value]?.nickname ?: "",
                            role = member.role
                        )
                    }
                )
            }
        )
    }

    override fun getTeamIdsOfUser(userId: UserId): List<Long> {
        val teams = teamService.teamsOfUser(userId)
        return teams.map { it.id().value }
    }

    override fun isUserLeaderOfTeam(userId: UserId, teamId: Long): Boolean {
        val team = teamService.activeTeam(TeamId.of(teamId))
        return teamService.membersOf(team)
            .filter { it.userId() == userId }
            .any { it.role == MemberRole.LEADER }
    }

    override fun isMemberOfTeam(userId: UserId, teamId: Long): Boolean {
        val team = teamService.activeTeam(TeamId.of(teamId))
        return teamService.membersOf(team).any { it.userId() == userId }
    }

    override fun getNicknameOfMember(memberId: Long): String {
        val member = teamService.memberOf(MemberId(memberId))
        return userInfoPort.findById(member.userId())?.nickname ?: ""
    }
}
